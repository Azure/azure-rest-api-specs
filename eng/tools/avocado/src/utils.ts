import * as YAML from 'js-yaml'

export const load = (content: string) => {
  try {
    return YAML.load(content) as any
  } catch (err) {
    console.log(err)
    return undefined
  }
}
