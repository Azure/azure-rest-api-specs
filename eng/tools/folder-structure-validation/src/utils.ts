import { access, readFile } from "fs/promises";
import defaultPath, { join, PlatformPath } from "path";

export async function fileExists(file: string) {
  return access(file)
    .then(() => true)
    .catch(() => false);
}

export async function readTspConfig(folder: string) {
  return readFile(join(folder, "tspconfig.yaml"), "utf-8");
}

export function normalizePath(folder: string, path: PlatformPath = defaultPath) {
  return normalizePathImpl(folder, path);
}

export function normalizePathImpl(folder: string, path: PlatformPath = defaultPath) {
  return path
    .resolve(folder)
    .split(path.sep)
    .join("/")
    .replace(/^([a-z]):/, (_match, driveLetter) => driveLetter.toUpperCase() + ":");
}
