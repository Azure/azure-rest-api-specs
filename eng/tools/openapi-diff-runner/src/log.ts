import { appendFileSync } from "fs";

/**
 * Log prefix for all messages from openapi-diff-runner
 */
export const LOG_PREFIX = "Runner-";

export enum LogLevel {
  Error = "error",
  Warn = "warn",
  Info = "info",
  Debug = "debug",
  Notice = "notice",
  Group = "group",
  EndGroup = "endgroup",
}

/**
 * Logs a message to the console with GitHub Actions workflow commands.
 * Automatically prefixes messages with LOG_PREFIX.
 * @param message The message to log.
 * @param level The log level (e.g., LogLevel.Group, LogLevel.EndGroup, LogLevel.Debug, LogLevel.Error).
 */
export function logMessage(message: string, level?: LogLevel): void {
  switch (level) {
    case LogLevel.Group: {
      console.log(`::group::${message}`);
      break;
    }
    case LogLevel.EndGroup: {
      console.log(`::endgroup::`);
      break;
    }
    case LogLevel.Debug: {
      console.log(`::debug::${message}`);
      break;
    }
    case LogLevel.Error: {
      console.log(`::error::${message}`);
      break;
    }
    case LogLevel.Warn: {
      console.log(`::warning::${message}`);
      break;
    }
    case LogLevel.Notice: {
      console.log(`::notice::${message}`);
      break;
    }
    case LogLevel.Info:
    default: {
      console.log(message);
      break;
    }
  }
}

/**
 * Log an error with file location information for GitHub Actions
 * Automatically prefixes messages with LOG_PREFIX.
 * @param message Error message
 * @param file File path (optional)
 * @param line Line number (optional)
 * @param col Column number (optional)
 */
export function logError(message: string, file?: string, line?: number, col?: number): void {
  if (file) {
    const location = line && col ? `line=${line},col=${col}` : line ? `line=${line}` : "";
    const fileLocation = location ? `file=${file},${location}` : `file=${file}`;
    console.log(`::error ${fileLocation}::${message}`);
  } else {
    console.log(`::error::${message}`);
  }
}

/**
 * Log a warning with file location information for GitHub Actions
 * Automatically prefixes messages with LOG_PREFIX.
 * @param message Warning message
 * @param file File path (optional)
 * @param line Line number (optional)
 * @param col Column number (optional)
 */
export function logWarning(message: string, file?: string, line?: number, col?: number): void {
  if (file) {
    const location = line && col ? `line=${line},col=${col}` : line ? `line=${line}` : "";
    const fileLocation = location ? `file=${file},${location}` : `file=${file}`;
    console.log(`::warning ${fileLocation}::${message}`);
  } else {
    console.log(`::warning::${message}`);
  }
}

/**
 * Set an output parameter in GitHub Actions
 * @param name Output parameter name
 * @param value Output parameter value
 */
export function setOutput(name: string, value: string): void {
  if (process.env.GITHUB_OUTPUT) {
    appendFileSync(process.env.GITHUB_OUTPUT, `${name}=${value}\n`);
  } else {
    // Fallback to older syntax
    console.log(`::set-output name=${name}::${value}`);
  }
}

/**
 * Add content to the GitHub Actions job summary
 * @param content Content to add to summary
 */
export function addToSummary(content: string): void {
  if (process.env.GITHUB_STEP_SUMMARY) {
    appendFileSync(process.env.GITHUB_STEP_SUMMARY, content);
  }
  // Do nothing if GITHUB_STEP_SUMMARY is not available
}

/**
 * Create a collapsible group in logs
 * @param title Group title
 * @param content Function that logs the group content
 */
export async function logGroup<T>(title: string, content: () => Promise<T> | T): Promise<T> {
  logMessage(title, LogLevel.Group);
  try {
    const result = await content();
    return result;
  } finally {
    logMessage("", LogLevel.EndGroup);
  }
}
