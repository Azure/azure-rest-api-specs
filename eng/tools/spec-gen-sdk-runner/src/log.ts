export enum LogLevel {
  Error = "error",
  Warn = "warn",
  Info = "info",
  Debug = "debug",
  Group = "group",
  EndGroup = "endgroup",
}

export enum LogIssueType {
  Error = "error",
  Warning = "warning",
}

/**
 * Logs a message to the console with a specified log level. *
 * @param message The message to log.
 * @param level The log level (e.g., LogLevel.Group, LogLevel.EndGroup, LogLevel.Debug, LogLevel.Error).
 */
export function logMessage(message: string, level?: LogLevel): void {
  switch (level) {
    case LogLevel.Group: {
      console.log(`##[group]${message}`);
      break;
    }
    case LogLevel.EndGroup: {
      console.log(`##[endgroup]`);
      break;
    }
    case LogLevel.Debug: {
      console.log(`[debug]${message}`);
      break;
    }
    case LogLevel.Error: {
      console.error(message);
      break;
    }
    case LogLevel.Warn: {
      console.warn(message);
      break;
    }
    default: {
      console.log(message);
      break;
    }
  }
}

export function vsoAddAttachment(name: string, path: string): void {
  console.log(`##vso[task.addattachment type=Distributedtask.Core.Summary;name=${name};]${path}`);
}

export function vsoLogIssue(message: string, type: LogIssueType = LogIssueType.Error): void {
  console.log(`##vso[task.logissue type=${type}]${message}`);
}

export function setVsoVariable(variable: string, value: string, isOutput = false): void {
  console.log(
    `##vso[task.setVariable variable=${variable}${isOutput ? ";isoutput=true" : ""}]${value}`,
  );
}
