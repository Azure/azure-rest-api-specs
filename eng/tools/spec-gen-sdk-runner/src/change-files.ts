import path from "node:path";
import {
  getChangedFiles,
  searchRelatedParentFolders,
  searchSharedLibrary,
  searchRelatedTypeSpecProjectBySharedLibrary,
  groupPathsByService,
  createCombinedSpecs,
  type SpecResults,
  type ChangedSpecs,
  getLastPathSegment,
} from "./utils.js";
import { logMessage } from "./log.js";
import { SpecGenSdkCmdInput } from "./types.js";

export const readmeMdRegex = /^readme.md$/;
export const typespecProjectRegex = /^tspconfig.yaml$/;
export const typespecProjectSharedLibraryRegex = /[^/]+\.Shared/;

export function detectChangedSpecConfigFiles(commandInput: SpecGenSdkCmdInput): ChangedSpecs[] {
  const prChangedFiles: string[] = getChangedFiles(commandInput.localSpecRepoPath) ?? [];
  if (prChangedFiles.length === 0) {
    logMessage("No files changed in the PR");
  }
  logMessage(`Changed files in the PR: ${prChangedFiles.length}`);
  for (const file of prChangedFiles) {
    logMessage(`\t${file}`);
  }
  const fileList = prChangedFiles
    .filter((p) => p.startsWith("specification/"))
    .filter((p) => !p.includes("/scenarios/"));

  logMessage(`Related readme.md and typespec project list:`);
  const changedSpecs: ChangedSpecs[] = [];

  const readmeMDResult = searchRelatedParentFolders(fileList, {
    searchFileRegex: readmeMdRegex,
    specRepoFolder: commandInput.localSpecRepoPath,
    stopAtFolder: "specification",
  });

  const typespecProjectResult = searchRelatedParentFolders(fileList, {
    searchFileRegex: typespecProjectRegex,
    specRepoFolder: commandInput.localSpecRepoPath,
    stopAtFolder: "specification",
  });

  const typespecProjectSharedLibraries = searchSharedLibrary(fileList, {
    searchFileRegex: typespecProjectSharedLibraryRegex,
    specRepoFolder: commandInput.localSpecRepoPath,
  });

  const typespecProjectResultSearchedBySharedLibrary = searchRelatedTypeSpecProjectBySharedLibrary(
    typespecProjectSharedLibraries,
    {
      searchFileRegex: typespecProjectRegex,
      specRepoFolder: commandInput.localSpecRepoPath,
    },
  );

  // Merge typespec project results
  for (const folderPath of Object.keys(typespecProjectResultSearchedBySharedLibrary)) {
    if (typespecProjectResult[folderPath]) {
      typespecProjectResult[folderPath] = [
        ...typespecProjectResult[folderPath],
        ...typespecProjectResultSearchedBySharedLibrary[folderPath],
      ];
    } else {
      typespecProjectResult[folderPath] = typespecProjectResultSearchedBySharedLibrary[folderPath];
    }
  }

  // Group paths by service
  const serviceMap = groupPathsByService(readmeMDResult, typespecProjectResult);

  const results: SpecResults = { readmeMDResult, typespecProjectResult };

  // Process each service
  for (const [, info] of serviceMap) {
    // Case: Resource Manager with .Management
    if (info.managementPaths.length > 0) {
      if (info.resourceManagerPaths.length === 1) {
        // Single resource-manager path - match with all Management paths
        const newSpecs = createCombinedSpecs(
          info.resourceManagerPaths[0].path,
          info.managementPaths,
          results,
        );
        changedSpecs.push(...newSpecs);
        logMessage(
          `\t readme folders: ${info.resourceManagerPaths[0].path}, tspconfig folders: ${info.managementPaths}`,
        );
        for (const p of info.managementPaths) {
          delete typespecProjectResult[p];
        }
        delete readmeMDResult[info.resourceManagerPaths[0].path];
      } else {
        // Multiple resource-manager paths - match by subfolder name
        for (const rmPath of info.resourceManagerPaths) {
          const matchingManagements = info.managementPaths.filter((mPath) => {
            const rmSubPath = rmPath.subPath;
            const managementName = getLastPathSegment(mPath).replace(".Management", "");
            return rmSubPath && rmSubPath === managementName;
          });
          if (matchingManagements.length > 0) {
            const newSpecs = createCombinedSpecs(rmPath.path, matchingManagements, results);
            changedSpecs.push(...newSpecs);
            logMessage(
              `\t readme folders: ${rmPath.path}, tspconfig folders: ${matchingManagements}`,
            );
            for (const p of matchingManagements) {
              delete typespecProjectResult[p];
            }
            delete readmeMDResult[rmPath.path];
          }
        }
      }
    }

    // Case: Data Plane matching
    if (info.dataPlanePaths.length > 0 && info.otherTypeSpecPaths.length > 0) {
      if (info.dataPlanePaths.length === 1) {
        // Single data-plane path - match with all non-Management TypeSpec paths
        const newSpecs = createCombinedSpecs(
          info.dataPlanePaths[0].path,
          info.otherTypeSpecPaths,
          results,
        );
        changedSpecs.push(...newSpecs);
        logMessage(
          `\t readme folders: ${info.dataPlanePaths[0].path}, tspconfig folders: ${info.otherTypeSpecPaths}`,
        );
        for (const p of info.otherTypeSpecPaths) {
          delete typespecProjectResult[p];
        }
        delete readmeMDResult[info.dataPlanePaths[0].path];
      } else {
        // Multiple data-plane paths - match by subfolder name
        for (const dpPath of info.dataPlanePaths) {
          const matchingTypeSpecs = info.otherTypeSpecPaths.filter((tsPath) => {
            const dpSubFolder = dpPath.subFolder;
            const tsLastSegment = getLastPathSegment(tsPath);
            return dpSubFolder && dpSubFolder === tsLastSegment;
          });
          if (matchingTypeSpecs.length > 0) {
            const newSpecs = createCombinedSpecs(dpPath.path, matchingTypeSpecs, results);
            changedSpecs.push(...newSpecs);
            logMessage(
              `\t readme folders: ${dpPath.path}, tspconfig folders: ${matchingTypeSpecs}`,
            );
            for (const p of matchingTypeSpecs) {
              delete typespecProjectResult[p];
            }
            delete readmeMDResult[dpPath.path];
          }
        }
      }
    }
  }

  // Process remaining unmatched paths
  for (const folderPath of new Set([
    ...Object.keys(readmeMDResult),
    ...Object.keys(typespecProjectResult),
  ])) {
    const cs: ChangedSpecs = {
      specs: [],
    };

    if (typespecProjectResult[folderPath]) {
      cs.specs = typespecProjectResult[folderPath];
      cs.typespecProject = path.join(folderPath, "tspconfig.yaml");
      logMessage(`\t tspconfig: ${cs.typespecProject}`);
    } else {
      cs.readmeMd = path.join(folderPath, "readme.md");
      cs.specs = readmeMDResult[folderPath];
      logMessage(`\t readme: ${cs.readmeMd}`);
    }

    changedSpecs.push(cs);
  }

  return changedSpecs;
}
