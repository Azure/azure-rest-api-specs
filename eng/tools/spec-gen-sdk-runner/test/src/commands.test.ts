import { describe, test, expect, vi, beforeEach } from "vitest";
import * as log from "../../src/log.js";
import * as utils from "../../src/utils.js";
import * as changeFiles from "../../src/change-files.js";
import { fileURLToPath } from "node:url";
import path from "node:path";
import fs from "node:fs";
import {
  setPipelineVariables,
  parseArguments,
  prepareSpecGenSdkCommand,
  getSpecPaths,
  logIssuesToPipeline,
  getBreakingChangeInfo,
  processBreakingChangeLabelArtifacts,
  generateSdkForSingleSpec,
  generateSdkForSpecPr,
  //   generateSdkForBatchSpecs,
} from "../../src/commands.js";
import * as commands from "../../src/commands.js";

// Get the absolute path to the repo root
const currentFilePath = fileURLToPath(import.meta.url);
const repoRoot = path.resolve(path.dirname(currentFilePath), "../fixtures/");

describe("commands.ts", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe("setPipelineVariables", () => {
    test("should set pipeline variables correctly", () => {
      vi.spyOn(log, "setVsoVariable").mockImplementation(() => {});

      setPipelineVariables(
        "sdk/security/keyvault/azcertificates",
        "Configurations: 'specification/contosowidgetmanager/resource-manager/readme.md', and CommitSHA: 'aaabbb', in SpecRepo: 'https://github.com/Azure/azure-rest-api-specs'",
      );

      expect(log.setVsoVariable).toHaveBeenCalledWith(
        "PrBranch",
        "sdkauto/sdk_security/keyvault/azcertificates",
      );
      expect(log.setVsoVariable).toHaveBeenCalledWith(
        "PrTitle",
        "[AutoPR sdk/security/keyvault/azcertificates]",
      );
      expect(log.setVsoVariable).toHaveBeenCalledWith(
        "PrBody",
        "Configurations: 'specification/contosowidgetmanager/resource-manager/readme.md', and CommitSHA: 'aaabbb', in SpecRepo: 'https://github.com/Azure/azure-rest-api-specs'",
      );
    });
  });

  describe("parseArguments", () => {
    test("runMode is release when it has no batch-type and no pr-number", () => {
      const mockArgs = [
        "--scp",
        `${repoRoot}specification/contosowidgetmanager/resource-manager/Microsoft.Contoso/preview/2021-10-01-preview/examples/Employees_Get.json"`,
        "--lang",
        "azure-sdk-for-go",
      ];
      vi.spyOn(process, "argv", "get").mockReturnValue(["node", "script", ...mockArgs]);
      const result = parseArguments();
      expect(result.localSpecRepoPath).toBe(
        `${repoRoot}specification/contosowidgetmanager/resource-manager/Microsoft.Contoso/preview/2021-10-01-preview/examples/Employees_Get.json"`,
      );
      expect(result.sdkRepoName).toBe("azure-sdk-for-go");
      expect(result.prNumber).toBe("");
      expect(result.runMode).toBe("release");
    });

    test("runMode is release when it has pr-number", () => {
      const mockArgs = [
        "--scp",
        `${repoRoot}specification/contosowidgetmanager/resource-manager/Microsoft.Contoso/preview/2021-10-01-preview/examples/Employees_Get.json"`,
        "--pr-number",
        "1234",
      ];
      vi.spyOn(process, "argv", "get").mockReturnValue(["node", "script", ...mockArgs]);

      const result = parseArguments();

      expect(result.localSpecRepoPath).toBe(
        `${repoRoot}specification/contosowidgetmanager/resource-manager/Microsoft.Contoso/preview/2021-10-01-preview/examples/Employees_Get.json"`,
      );
      expect(result.sdkRepoName).toBe("azure-sdk-for-net");
      expect(result.prNumber).toBe("1234");
      expect(result.runMode).toBe("spec-pull-request");
    });

    test("runMode is batch when it has batch-type", () => {
      const mockArgs = [
        "--scp",
        `${repoRoot}specification/contosowidgetmanager/resource-manager/Microsoft.Contoso/preview/2021-10-01-preview/examples/Employees_Get.json"`,
        "--batch-type",
        "all-specs",
      ];
      vi.spyOn(process, "argv", "get").mockReturnValue(["node", "script", ...mockArgs]);

      const result = parseArguments();

      expect(result.localSpecRepoPath).toBe(
        `${repoRoot}specification/contosowidgetmanager/resource-manager/Microsoft.Contoso/preview/2021-10-01-preview/examples/Employees_Get.json"`,
      );
      expect(result.sdkRepoName).toBe("azure-sdk-for-net");
      expect(result.prNumber).toBe("");
      expect(result.runMode).toBe("batch");
    });
  });

  describe("prepareSpecGenSdkCommand", () => {
    test("should prepare the command correctly", () => {
      const commandInput = {
        localSpecRepoPath: "/spec/path",
        localSdkRepoPath: "/sdk/path",
        workingFolder: "/working/folder",
        sdkRepoName: "azure-sdk-for-js",
        specCommitSha: "abc123",
        runMode: "release",
        specRepoHttpsUrl: "https://github.com/spec",
        prNumber: "123",
        tspConfigPath: "config/path",
        readmePath: "readme/path",
        headRepoHttpsUrl: "https://github.com/head",
        headBranch: "main",
        apiVersion: "2021-01-01",
        sdkReleaseType: "beta",
        sdkLanguage: "typescript",
      };

      const result = prepareSpecGenSdkCommand(commandInput);

      expect(result).toContain("spec-gen-sdk");
      expect(result).toContain("--scp");
      expect(result).toContain("/spec/path");
      expect(result).toContain("--pr-number");
      expect(result).toContain("123");
    });
  });

  describe("getSpecPaths", () => {
    test("should return all specs for 'all-specs' batch type", () => {
      vi.spyOn(utils, "getAllTypeSpecPaths").mockReturnValue(["typespec1", "typespec2"]);
      vi.spyOn(utils, "findReadmeFiles").mockReturnValue(["readme1", "readme2"]);

      const result = getSpecPaths("all-specs", "/spec/path");

      expect(result).toEqual(["typespec1", "typespec2", "readme1", "readme2"]);
    });

    test("should return only TypeSpec paths for 'all-typespecs' batch type", () => {
      vi.spyOn(utils, "getAllTypeSpecPaths").mockReturnValue(["typespec1", "typespec2"]);

      const result = getSpecPaths("all-typespecs", "/spec/path");

      expect(result).toEqual(["typespec1", "typespec2"]);
    });
  });

  describe("logIssuesToPipeline", () => {
    test("should log errors and warnings to pipeline", () => {
      const mockLogContent = {
        key1: { errors: ["error1"], warnings: ["warning1"] },
        key2: { errors: ["error2"], warnings: [] },
      };
      vi.spyOn(fs, "readFileSync").mockReturnValue(JSON.stringify(mockLogContent));
      vi.spyOn(log, "logMessage").mockImplementation(() => {});
      vi.spyOn(log, "vsoLogIssue").mockImplementation(() => {});

      logIssuesToPipeline("/log/path", "spec config");

      expect(log.logMessage).toHaveBeenCalledWith(
        "Errors occurred while generating SDK from spec config",
        "group",
      );
      expect(log.vsoLogIssue).toHaveBeenCalledWith(
        "Errors occurred while generating SDK from spec config%0D%0Aerror1%0D%0Aerror2",
      );
    });
  });

  describe("getBreakingChangeInfo", () => {
    test("should return breaking change info if applicable", () => {
      const mockExecutionReport = {
        packages: [{ shouldLabelBreakingChange: true, breakingChangeLabel: "breaking-change" }],
      };

      const result = getBreakingChangeInfo(mockExecutionReport);

      expect(result).toEqual([true, "breaking-change"]);
    });

    test("should return no breaking change info if not applicable", () => {
      const mockExecutionReport = {
        packages: [{ shouldLabelBreakingChange: false, breakingChangeLabel: "" }],
      };

      const result = getBreakingChangeInfo(mockExecutionReport);

      expect(result).toEqual([false, ""]);
    });
  });

  describe("processBreakingChangeLabelArtifacts", () => {
    beforeEach(() => {
      vi.clearAllMocks();
    });
    test("should process breaking change label artifacts successfully", () => {
      vi.spyOn(fs, "existsSync").mockReturnValue(false);
      vi.spyOn(fs, "mkdirSync").mockImplementation(() => undefined);
      vi.spyOn(fs, "writeFileSync").mockImplementation(() => {});
      vi.spyOn(log, "setVsoVariable").mockImplementation(() => {});

      const mockCommandInput = {
        workingFolder: "/working/folder",
        sdkLanguage: "javascript",
        runMode: "",
        localSpecRepoPath: "",
        localSdkRepoPath: "",
        sdkRepoName: "",
        specCommitSha: "",
        specRepoHttpsUrl: "",
      };

      const result = processBreakingChangeLabelArtifacts(mockCommandInput, true, "breaking-change");

      const breakingChangeLabelArtifactPath = "/working/folder/out/breaking-change-label-artifact";
      expect(result).toBe(0);
      expect(fs.mkdirSync).toHaveBeenCalledWith(breakingChangeLabelArtifactPath, {
        recursive: true,
      });
      expect(fs.writeFileSync).toHaveBeenCalledWith(
        path.join(breakingChangeLabelArtifactPath, "spec-gen-sdk-breaking-change-artifact.json"),
        JSON.stringify({
          language: "javascript",
          labelAction: true,
        }),
      );
      expect(log.setVsoVariable).toHaveBeenCalledWith(
        "BreakingChangeLabelArtifactName",
        "spec-gen-sdk-breaking-change-artifact",
      );
      expect(log.setVsoVariable).toHaveBeenCalledWith(
        "BreakingChangeLabelArtifactPath",
        "out/breaking-change-label-artifact",
      );
      expect(log.setVsoVariable).toHaveBeenCalledWith("BreakingChangeLabelAction", "add");
      expect(log.setVsoVariable).toHaveBeenCalledWith("BreakingChangeLabel", "breaking-change");
    });

    test("should handle errors during artifact processing", () => {
      vi.spyOn(fs, "existsSync").mockReturnValue(false);
      vi.spyOn(fs, "mkdirSync").mockImplementation(() => {
        throw new Error("mkdir failed");
      });
      vi.spyOn(log, "logMessage").mockImplementation(() => {});
      vi.spyOn(log, "vsoLogIssue").mockImplementation(() => {});

      const mockCommandInput = {
        workingFolder: "/working/folder",
        sdkLanguage: "javascript",
        runMode: "",
        localSpecRepoPath: "",
        localSdkRepoPath: "",
        sdkRepoName: "",
        specCommitSha: "",
        specRepoHttpsUrl: "",
      };
      const result = processBreakingChangeLabelArtifacts(mockCommandInput, true, "breaking-change");

      expect(result).toBe(1);
      expect(log.logMessage).toHaveBeenCalledWith("ending group logging", "endgroup");
    });
  });

  describe("generateSdkForSingleSpec", () => {
    beforeEach(() => {
      vi.clearAllMocks();
    });

    test("should generate SDK successfully", async () => {
      vi.spyOn(fs, "readFileSync").mockReturnValue(
        JSON.stringify({
          executionResult: "succeeded",
          packages: [{ packageName: "test-package", installationInstructions: "install" }],
        }),
      );
      vi.spyOn(log, "setVsoVariable").mockImplementation(() => {});
      vi.spyOn(utils, "runSpecGenSdkCommand").mockResolvedValueOnce(undefined);
      const result = await generateSdkForSingleSpec();

      expect(result).toBe(0);
      expect(utils.runSpecGenSdkCommand).toHaveBeenCalled();
      expect(log.setVsoVariable).toHaveBeenCalledWith("PrBranch", "sdkauto/test-package");
    });

    test("should handle errors during SDK generation", async () => {
      vi.spyOn(utils, "runSpecGenSdkCommand").mockRejectedValueOnce(new Error("Command failed"));
      const result = await generateSdkForSingleSpec();

      expect(result).toBe(1);
      expect(utils.runSpecGenSdkCommand).toHaveBeenCalled();
    });
  });

  describe("generateSdkForSpecPr", () => {
    beforeEach(() => {
      vi.clearAllMocks();
    });

    test("should process changed specs and generate SDKs", async () => {
      vi.spyOn(changeFiles, "detectChangedSpecConfigFiles").mockReturnValueOnce([
        {
          specs: [
            "specification/contosowidgetmanager/data-plane/Azure.Contoso.WidgetManager/preview/2022-11-01-preview/widgets.json",
            "specification/contosowidgetmanager/Contoso.WidgetManager/tspconfig.yaml",
            "specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.tsp",
          ],
          readmeMd: "specification/contosowidgetmanager/data-plane/readme.md",
          typespecProject:
            "specification/contosowidgetmanager/Contoso.WidgetManager/tspconfig.yaml",
        },
      ]);
      vi.spyOn(utils, "resetGitRepo").mockResolvedValueOnce(undefined);
      vi.spyOn(utils, "runSpecGenSdkCommand").mockResolvedValueOnce(undefined);
      vi.spyOn(fs, "readFileSync").mockReturnValue(
        JSON.stringify({
          executionResult: "succeeded",
          packages: [{ shouldLabelBreakingChange: true, breakingChangeLabel: "breaking-change" }],
        }),
      );

      vi.spyOn(fs, "existsSync").mockReturnValue(false);
      vi.spyOn(fs, "mkdirSync").mockImplementation(() => undefined);
      vi.spyOn(fs, "writeFileSync").mockImplementation(() => {});

      const result = await generateSdkForSpecPr();
      expect(utils.runSpecGenSdkCommand).toHaveBeenCalled();
      expect(result).toBe(0);
    });

    test("should handle errors during spec processing", async () => {
      vi.spyOn(changeFiles, "detectChangedSpecConfigFiles").mockReturnValueOnce([
        {
          specs: [
            "specification/contosowidgetmanager/data-plane/Azure.Contoso.WidgetManager/preview/2022-11-01-preview/widgets.json",
            "specification/contosowidgetmanager/Contoso.WidgetManager/tspconfig.yaml",
            "specification/contosowidgetmanager/Contoso.WidgetManager.Shared/main.tsp",
          ],
          readmeMd: "specification/contosowidgetmanager/data-plane/readme.md",
          typespecProject:
            "specification/contosowidgetmanager/Contoso.WidgetManager/tspconfig.yaml",
        },
      ]);
      vi.spyOn(utils, "runSpecGenSdkCommand").mockRejectedValueOnce(new Error("Command failed"));

      const result = await generateSdkForSpecPr();

      expect(result).toBe(1);
    });
  });

  describe("generateSdkForBatchSpecs", () => {
    beforeEach(() => {
      vi.clearAllMocks();
    });

    test("should generate SDKs for batch specs successfully", async () => {
      // Mock getSpecPaths to return a list of specs
      vi.spyOn(commands, "getSpecPaths").mockReturnValueOnce(["spec1", "spec2"]);

      // Mock other dependencies
      vi.spyOn(utils, "resetGitRepo").mockResolvedValueOnce(undefined);
      vi.spyOn(utils, "runSpecGenSdkCommand").mockResolvedValue(undefined);
      vi.spyOn(fs, "readFileSync").mockReturnValue(
        JSON.stringify({
          executionResult: "succeeded",
        }),
      );

      vi.spyOn(log, "logMessage").mockImplementation(() => {});

      // Call the function
      const result = await commands.generateSdkForBatchSpecs("all-specs");

      // Assertions
      expect(result).toBe(0);
      expect(commands.getSpecPaths).toHaveBeenCalledWith("all-specs", expect.any(String));
      expect(utils.runSpecGenSdkCommand).toHaveBeenCalledTimes(2); // Called for each spec
      expect(log.logMessage).toHaveBeenCalledWith("Runner command executed successfully");
    });

    /**
    test("should handle errors during batch spec generation", async () => {
      // Mock getSpecPaths to return a list of specs
      vi.spyOn(commands, "getSpecPaths").mockReturnValueOnce(["spec1", "spec2"]);
  
      // Mock runSpecGenSdkCommand to throw an error for the first spec
      vi.spyOn(utils, "runSpecGenSdkCommand")
        .mockRejectedValueOnce(new Error("Command failed"))
        .mockResolvedValueOnce(undefined); // Second spec succeeds
  
      // Mock other dependencies
      vi.spyOn(utils, "resetGitRepo").mockResolvedValueOnce(undefined);
      vi.spyOn(commands, "getExecutionReport").mockReturnValue({
        executionResult: "succeeded",
      });
      vi.spyOn(log, "logMessage").mockImplementation(() => {});
  
      // Call the function
      const result = await commands.generateSdkForBatchSpecs("all-specs");
  
      // Assertions
      expect(result).toBe(1); // Overall result should indicate failure
      expect(commands.getSpecPaths).toHaveBeenCalledWith("all-specs", expect.any(String));
      expect(utils.runSpecGenSdkCommand).toHaveBeenCalledTimes(2); // Called for each spec
      expect(log.logMessage).toHaveBeenCalledWith(
        "Runner: error executing command:Error: Command failed",
        "error"
      );
    });
    test("should handle empty spec paths gracefully", async () => {
      vi.spyOn(commands, "getSpecPaths").mockReturnValueOnce([]);
  
      vi.spyOn(log, "logMessage").mockImplementation(() => {});
  
      const result = await commands.generateSdkForBatchSpecs("all-specs");
  
      expect(result).toBe(0); // No specs to process, so result should be success
      expect(commands.getSpecPaths).toHaveBeenCalledWith("all-specs", expect.any(String));
      expect(utils.runSpecGenSdkCommand).not.toHaveBeenCalled(); // No specs, so no commands run
      expect(log.logMessage).toHaveBeenCalledWith("No specs to process for batch type: all-specs");
    });

     */
  });
});
