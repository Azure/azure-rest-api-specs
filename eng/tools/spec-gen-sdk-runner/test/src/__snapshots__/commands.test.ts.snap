// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateSdkForBatchSpecs > should generate SDKs for all specs successfully 1`] = `
[
  [
    "/working/folder/out/logs/generation-summary.md",
    "
## Successful Specs in the Generation Process
typespec1,typespec2,readme1,readme2,
## Total Successful Specs
 4
## Total Specs Count
 4

",
  ],
]
`;

exports[`generateSdkForBatchSpecs > should handle empty spec paths gracefully 1`] = `
[
  [
    "/working/folder/out/logs/generation-summary.md",
    "
## Total Specs Count
 0

",
  ],
]
`;

exports[`generateSdkForBatchSpecs > should handle errors during SDK generation 1`] = `
[
  [
    "/working/folder/out/logs/generation-summary.md",
    "
## Spec Failures in the Generation Process
typespec1,typespec2,
## Total Failed Specs
 2
## Total Specs Count
 2

",
  ],
]
`;
